### ISSUE TITLE
injectable 2.0.0 registers in wrong order for abstracts and generic type

### ISSUE THREAD

**Issue by naamapps** [Milad-Akarie/injectable#324]
After upgrading to latest version, and running the build runner cmd, the order of registrations is completely wrong and causes the loading of the app to fail.

I don't know exactly what is the issue, but with previous version everything worked great.

Please advise,
Thanks

**Comment by Milad-Akarie** [comment]
Hello @naamapps I double checked and re-run order tests but everything works as expected. 
can you please provide more details? a git comparison maybe?

**Comment by naamapps** [comment]
See the difference, the problem is mostly with the ListServiceBase that is registered after it is used by SearchService for example.

Old version:

```dart

// ignore_for_file: lines_longer_than_80_chars
/// initializes the registration of provided dependencies inside of [GetIt]
Future<_i1.GetIt> $initGetIt(
  _i1.GetIt get, {
  String? environment,
  _i2.EnvironmentFilter? environmentFilter,
}) async {
  final gh = _i2.GetItHelper(
    get,
    environment,
    environmentFilter,
  );
  final registerModule = _$RegisterModule();
  final poolManagerModule = _$PoolManagerModule();
  gh.singleton<_i3.AppNavigationService>(_i3.AppNavigationService());
  gh.lazySingleton<_i4.BaseRepository<_i5.LocalDbItem, _i6.User>>(
    () => _i7.UsersSearchHistoryRepository(),
    instanceName: 'UsersSearchHistoryRepository',
  );
  gh.lazySingleton<_i4.BaseRepository<_i5.LocalDbItem, _i8.Post>>(
    () => _i9.DraftsRepository(),
    instanceName: 'DraftsRepository',
  );
  gh.lazySingleton<_i4.BaseRepository<_i5.LocalDbItem, _i8.Post>>(
    () => _i10.PostsSearchHistoryRepository(),
    instanceName: 'PostsSearchHistoryRepository',
  );
  await gh.singletonAsync<_i11.DeviceInfo>(
    () => registerModule.deviceInfo,
    preResolve: true,
  );
  gh.lazySingleton<_i12.DialogService>(() => _i12.DialogService());
  gh.singleton<_i13.FlutterSecureStorage>(registerModule.flutterSecureStorage);
  gh.lazySingleton<_i14.GlobalStateService>(() => _i14.GlobalStateService());
  gh.singleton<_i15.HttpService>(_i15.HttpService(get<_i12.DialogService>()));
  gh.lazySingleton<_i16.KeyboardManager>(() => _i16.KeyboardManager());
  gh.lazySingleton<_i17.MentionsService>(() => _i17.MentionsService());
  gh.factory<_i18.NotificationService>(() => _i18.NotificationService());
  gh.singleton<_i19.NotificationsService>(
      _i19.NotificationsService(get<_i15.HttpService>()));
  await gh.singletonAsync<_i20.PackageInfo>(
    () => registerModule.packageInfo,
    preResolve: true,
  );
  gh.lazySingleton<_i21.PoolManager<String, _i18.NotificationService>>(
      () => poolManagerModule.notificationServicePoolManager(
            get<_i18.NotificationService>(),
            get<_i14.GlobalStateService>(),
          ));
  gh.lazySingleton<_i22.PostDraftsService>(() => _i22.PostDraftsService(
        get<_i12.DialogService>(),
        get<_i4.BaseRepository<_i5.LocalDbItem, _i8.Post>>(
            instanceName: 'DraftsRepository'),
      ));
  gh.singleton<_i23.PostsService>(_i23.PostsService(
    get<_i17.MentionsService>(),
    get<_i15.HttpService>(),
  ));
  await gh.singletonAsync<_i24.QuickActionsService>(
    () => _i24.QuickActionsService.getInstance(get<_i3.AppNavigationService>()),
    preResolve: true,
  );
  await gh.singletonAsync<_i25.SharedPreferences>(
    () => registerModule.sharedPreferences,
    preResolve: true,
  );
  gh.singleton<_i26.SpeechToTextService>(_i26.SpeechToTextService());
  gh.lazySingleton<_i27.ThemeService>(() => registerModule.themeService);
  gh.singleton<_i28.TrackingService>(_i28.TrackingService());
  await gh.singletonAsync<_i29.AppConfig>(
    () => registerModule.appConfig(
      get<_i25.SharedPreferences>(),
      get<_i20.PackageInfo>(),
    ),
    preResolve: true,
  );
  await gh.singletonAsync<_i30.AuthService>(
    () => _i30.AuthService.getInstance(
      get<_i15.HttpService>(),
      get<_i14.GlobalStateService>(),
      get<_i13.FlutterSecureStorage>(),
    ),
    preResolve: true,
  );
  gh.lazySingleton<_i31.CommentsService>(
      () => _i31.CommentsService(get<_i15.HttpService>()));
  gh.singleton<_i37.DioService>(_i37.DioService(
    get<_i30.AuthService>(),
    get<_i11.DeviceInfo>(),
    get<_i29.AppConfig>(),
  ));
  gh.factory<
      _i38.ListServiceBase<
          _i39.Notification<_i40.Serializable, _i40.Serializable>,
          _i18.NotificationService>>(() => _i41.NotificationsListService(
        get<_i21.PoolManager<String, _i18.NotificationService>>(),
        get<_i15.HttpService>(),
      ));
  gh.lazySingleton<_i42.ReportService>(
      () => _i42.ReportService(get<_i29.AppConfig>()));
  gh.singleton<_i43.SettingsViewModel>(_i43.SettingsViewModel(
    get<_i27.ThemeService>(),
    get<_i30.AuthService>(),
    get<_i12.DialogService>(),
    get<_i29.AppConfig>(),
  ));
  gh.lazySingleton<_i44.ShareService<_i8.Post>>(() => _i45.SharePostService(
        get<_i12.DialogService>(),
        get<_i29.AppConfig>(),
        get<_i15.HttpService>(),
      ));
  gh.lazySingleton<_i44.ShareService<_i6.User>>(() => _i46.ShareUserService(
        get<_i12.DialogService>(),
        get<_i29.AppConfig>(),
      ));
  gh.singleton<_i47.UsersService>(_i47.UsersService(
    get<_i15.HttpService>(),
    get<_i17.MentionsService>(),
    get<_i30.AuthService>(),
  ));
  gh.factory<_i48.ProfileService>(() => _i48.ProfileService(
        get<_i44.ShareService<_i6.User>>(),
        get<_i30.AuthService>(),
        get<_i47.UsersService>(),
        get<_i12.DialogService>(),
        get<_i14.GlobalStateService>(),
        get<_i42.ReportService>(),
      ));
  gh.lazySingleton<_i21.PoolManager<String, _i48.ProfileService>>(
      () => poolManagerModule.profileServicePoolManager(
            get<_i48.ProfileService>(),
            get<_i14.GlobalStateService>(),
          ));
  gh.factory<_i38.ListServiceBase<_i6.User, _i48.ProfileService>>(
      () => _i49.UsersListService(
            get<_i21.PoolManager<String, _i48.ProfileService>>(),
            get<_i15.HttpService>(),
          ));
  gh.lazySingleton<_i50.UsersSuggestionsViewModel>(
      () => _i50.UsersSuggestionsViewModel(
            get<_i38.ListServiceBase<_i6.User, _i48.ProfileService>>(),
            get<_i14.GlobalStateService>(),
          ));
  gh.factory<_i51.CommentService>(() => _i51.CommentService(
        get<_i30.AuthService>(),
        get<_i12.DialogService>(),
        get<_i42.ReportService>(),
        get<_i31.CommentsService>(),
        get<_i14.GlobalStateService>(),
        get<_i38.ListServiceBase<_i6.User, _i48.ProfileService>>(),
      ));
  gh.lazySingleton<_i21.PoolManager<String, _i51.CommentService>>(
      () => poolManagerModule.commentServicePoolManager(
            get<_i51.CommentService>(),
            get<_i14.GlobalStateService>(),
          ));
  gh.factory<_i38.ListServiceBase<_i52.Comment, _i51.CommentService>>(
      () => _i53.CommentsListService(
            get<_i21.PoolManager<String, _i51.CommentService>>(),
            get<_i15.HttpService>(),
            get<_i31.CommentsService>(),
            get<_i14.GlobalStateService>(),
            get<_i17.MentionsService>(),
            get<_i16.KeyboardManager>(),
          ));
  gh.factory<_i54.PostService>(() => _i54.PostService(
        get<_i3.AppNavigationService>(),
        get<_i23.PostsService>(),
        get<_i30.AuthService>(),
        get<_i44.ShareService<_i8.Post>>(),
        get<_i12.DialogService>(),
        get<_i14.GlobalStateService>(),
        get<_i42.ReportService>(),
        get<_i38.ListServiceBase<_i52.Comment, _i51.CommentService>>(),
        get<_i38.ListServiceBase<_i6.User, _i48.ProfileService>>(),
      ));
  gh.lazySingleton<_i21.PoolManager<String, _i54.PostService>>(
      () => poolManagerModule.postServicePoolManager(
            get<_i54.PostService>(),
            get<_i14.GlobalStateService>(),
          ));
  gh.factory<_i38.ListServiceBase<_i8.Post, _i54.PostService>>(
      () => _i55.PostsListService(
            get<_i14.GlobalStateService>(),
            get<_i21.PoolManager<String, _i54.PostService>>(),
            get<_i15.HttpService>(),
          ));
  await gh.singletonAsync<_i56.SearchService>(
    () => _i56.SearchService.getInstance(
      get<_i15.HttpService>(),
      get<_i38.ListServiceBase<_i8.Post, _i54.PostService>>(),
      get<_i38.ListServiceBase<_i6.User, _i48.ProfileService>>(),
      get<_i14.GlobalStateService>(),
    ),
    preResolve: true,
  );
  return get;
}

class _$RegisterModule extends _i57.RegisterModule {}

class _$PoolManagerModule extends _i21.PoolManagerModule {}

```

New Version:
```dart

/// ignore_for_file: unnecessary_lambdas
/// ignore_for_file: lines_longer_than_80_chars
extension GetItInjectableX on _i1.GetIt {
  /// initializes the registration of main-scope dependencies inside of [GetIt]
  Future<_i1.GetIt> init({
    String? environment,
    _i2.EnvironmentFilter? environmentFilter,
  }) async {
    final gh = _i2.GetItHelper(
      this,
      environment,
      environmentFilter,
    );
    final registerModule = _$RegisterModule();
    final poolManagerModule = _$PoolManagerModule();
    gh.singleton<_i3.PostsService>(_i3.PostsService(
      gh<_i4.MentionsService>(),
      gh<_i5.HttpService>(),
    ));
    gh.lazySingleton<_i6.BaseRepository<_i7.LocalDbItem, _i8.Post>>(
      () => _i9.PostsSearchHistoryRepository(),
      instanceName: 'PostsSearchHistoryRepository',
    );
    gh.lazySingleton<_i6.BaseRepository<_i7.LocalDbItem, _i10.User>>(
      () => _i11.UsersSearchHistoryRepository(),
      instanceName: 'UsersSearchHistoryRepository',
    );
    gh.lazySingleton<_i6.BaseRepository<_i7.LocalDbItem, _i8.Post>>(
      () => _i12.DraftsRepository(),
      instanceName: 'DraftsRepository',
    );
    await gh.singletonAsync<_i13.DeviceInfo>(
      () => registerModule.deviceInfo,
      preResolve: true,
    );
    gh.lazySingleton<_i14.DialogService>(() => _i14.DialogService());
    gh.singleton<_i15.FlutterSecureStorage>(
        registerModule.flutterSecureStorage);
    gh.lazySingleton<_i16.GlobalStateService>(() => _i16.GlobalStateService());
    gh.singleton<_i5.HttpService>(_i5.HttpService(gh<_i14.DialogService>()));
    gh.lazySingleton<_i17.KeyboardManager>(() => _i17.KeyboardManager());
    gh.lazySingleton<_i4.MentionsService>(() => _i4.MentionsService());
    gh.singleton<_i18.NotificationsService>(
        _i18.NotificationsService(gh<_i5.HttpService>()));
    await gh.singletonAsync<_i19.PackageInfo>(
      () => registerModule.packageInfo,
      preResolve: true,
    );
    gh.factory<
            _i20.Poolable<
                _i21.Notification<_i22.Serializable, _i22.Serializable>>>(
        () => _i23.NotificationService());
    gh.lazySingleton<_i24.PostDraftsService>(() => _i24.PostDraftsService(
          gh<_i14.DialogService>(),
          gh<_i6.BaseRepository<_i7.LocalDbItem, _i8.Post>>(
              instanceName: 'DraftsRepository'),
        ));
    gh.singleton<_i25.AppNavigationService>(_i25.AppNavigationService());
    await gh.singletonAsync<_i26.QuickActionsService>(
      () =>
          _i26.QuickActionsService.getInstance(gh<_i25.AppNavigationService>()),
      preResolve: true,
    );
    await gh.singletonAsync<_i27.SharedPreferences>(
      () => registerModule.sharedPreferences,
      preResolve: true,
    );
    gh.singleton<_i28.SpeechToTextService>(_i28.SpeechToTextService());
    gh.lazySingleton<_i29.ThemeService>(() => registerModule.themeService);
    gh.singleton<_i30.TrackingService>(_i30.TrackingService());
    await gh.singletonAsync<_i31.AppConfig>(
      () => registerModule.appConfig(
        gh<_i27.SharedPreferences>(),
        gh<_i19.PackageInfo>(),
      ),
      preResolve: true,
    );
    await gh.singletonAsync<_i32.AuthService>(
      () => _i32.AuthService.getInstance(
        gh<_i5.HttpService>(),
        gh<_i16.GlobalStateService>(),
        gh<_i15.FlutterSecureStorage>(),
      ),
      preResolve: true,
    );
    gh.lazySingleton<_i33.CommentsService>(
        () => _i33.CommentsService(gh<_i5.HttpService>()));
    gh.singleton<_i39.DioService>(_i39.DioService(
      gh<_i32.AuthService>(),
      gh<_i13.DeviceInfo>(),
      gh<_i31.AppConfig>(),
    ));
    gh.lazySingleton<
        _i20.PoolManager<
            String,
            _i20.Poolable<
                _i21.Notification<_i22.Serializable, _i22.Serializable>>>>(() =>
        poolManagerModule.notificationServicePoolManager(
          gh<
              _i20.Poolable<
                  _i21.Notification<_i22.Serializable, _i22.Serializable>>>(),
          gh<_i16.GlobalStateService>(),
        ));
    gh.lazySingleton<_i40.ReportService>(
        () => _i40.ReportService(gh<_i31.AppConfig>()));
    await gh.singletonAsync<_i41.SearchService>(
      () => _i41.SearchService.getInstance(
        gh<_i5.HttpService>(),
        gh<_i42.ListServiceBase<_i8.Post>>(),
        gh<_i42.ListServiceBase<_i10.User>>(),
        gh<_i16.GlobalStateService>(),
      ),
      preResolve: true,
    );
    gh.lazySingleton<_i43.ShareService<_i10.User>>(() => _i44.ShareUserService(
          gh<_i14.DialogService>(),
          gh<_i31.AppConfig>(),
        ));
    gh.lazySingleton<_i43.ShareService<_i8.Post>>(() => _i45.SharePostService(
          gh<_i14.DialogService>(),
          gh<_i31.AppConfig>(),
          gh<_i5.HttpService>(),
        ));
    gh.singleton<_i46.UsersService>(_i46.UsersService(
      gh<_i5.HttpService>(),
      gh<_i4.MentionsService>(),
      gh<_i32.AuthService>(),
    ));
    gh.factory<
            _i42.ListServiceBase<
                _i21.Notification<_i22.Serializable, _i22.Serializable>>>(
        () => _i47.NotificationsListService(
              gh<
                  _i20.PoolManager<
                      String,
                      _i20.Poolable<
                          _i21.Notification<_i22.Serializable,
                              _i22.Serializable>>>>(),
              gh<_i5.HttpService>(),
            ));
    gh.factory<_i20.Poolable<_i10.User>>(() => _i48.ProfileService(
          gh<_i43.ShareService<_i10.User>>(),
          gh<_i32.AuthService>(),
          gh<_i46.UsersService>(),
          gh<_i14.DialogService>(),
          gh<_i16.GlobalStateService>(),
          gh<_i40.ReportService>(),
        ));
    gh.lazySingleton<_i20.PoolManager<String, _i20.Poolable<_i10.User>>>(
        () => poolManagerModule.profileServicePoolManager(
              gh<_i20.Poolable<_i10.User>>(),
              gh<_i16.GlobalStateService>(),
            ));
    gh.factory<_i42.ListServiceBase<_i10.User>>(() => _i49.UsersListService(
          gh<_i20.PoolManager<String, _i20.Poolable<_i10.User>>>(),
          gh<_i5.HttpService>(),
        ));
    gh.factory<_i20.Poolable<_i50.Comment>>(() => _i51.CommentService(
          gh<_i32.AuthService>(),
          gh<_i14.DialogService>(),
          gh<_i40.ReportService>(),
          gh<_i33.CommentsService>(),
          gh<_i16.GlobalStateService>(),
          gh<_i42.ListServiceBase<_i10.User>>(),
        ));
    gh.lazySingleton<_i52.UsersSuggestionsViewModel>(
        () => _i52.UsersSuggestionsViewModel(
              gh<_i42.ListServiceBase<_i10.User>>(),
              gh<_i16.GlobalStateService>(),
            ));
    gh.lazySingleton<_i20.PoolManager<String, _i20.Poolable<_i50.Comment>>>(
        () => poolManagerModule.commentServicePoolManager(
              gh<_i20.Poolable<_i50.Comment>>(),
              gh<_i16.GlobalStateService>(),
            ));
    gh.factory<_i42.ListServiceBase<_i50.Comment>>(
        () => _i53.CommentsListService(
              gh<_i20.PoolManager<String, _i20.Poolable<_i50.Comment>>>(),
              gh<_i5.HttpService>(),
              gh<_i33.CommentsService>(),
              gh<_i16.GlobalStateService>(),
              gh<_i4.MentionsService>(),
              gh<_i17.KeyboardManager>(),
            ));
    gh.factory<_i20.Poolable<_i8.Post>>(() => _i54.PostService(
          gh<_i25.AppNavigationService>(),
          gh<_i3.PostsService>(),
          gh<_i32.AuthService>(),
          gh<_i43.ShareService<_i8.Post>>(),
          gh<_i14.DialogService>(),
          gh<_i16.GlobalStateService>(),
          gh<_i40.ReportService>(),
          gh<_i42.ListServiceBase<_i50.Comment>>(),
          gh<_i42.ListServiceBase<_i10.User>>(),
        ));
    gh.lazySingleton<_i20.PoolManager<String, _i20.Poolable<_i8.Post>>>(
        () => poolManagerModule.postServicePoolManager(
              gh<_i20.Poolable<_i8.Post>>(),
              gh<_i16.GlobalStateService>(),
            ));
    gh.factory<_i42.ListServiceBase<_i8.Post>>(() => _i55.PostsListService(
          gh<_i16.GlobalStateService>(),
          gh<_i20.PoolManager<String, _i20.Poolable<_i8.Post>>>(),
          gh<_i5.HttpService>(),
        ));
    gh.singleton<_i56.SettingsViewModel>(_i56.SettingsViewModel(
      gh<_i29.ThemeService>(),
      gh<_i32.AuthService>(),
      gh<_i14.DialogService>(),
      gh<_i31.AppConfig>(),
    ));
    return this;
  }
}

class _$RegisterModule extends _i57.RegisterModule {}

class _$PoolManagerModule extends _i20.PoolManagerModule {}

```

**Comment by Milad-Akarie** [comment]
@naamapps I found out that in my previous refactor I'm double sorting dependencies, I guess that's what was causing the issue. 
please confirm if the issue is fixed  in version 2.0.1 then close it.

**Comment by naamapps** [comment]
@Milad-Akarie unfortunately it did not solve the issue..
Still with wrong order generation

**Comment by Milad-Akarie** [comment]
@naamapps you used injectable_generator 2.0.1 right?

**Comment by naamapps** [comment]
@Milad-Akarie Of Course, Upgraded both injectable and injectable_generator to 2.0.1

**Comment by PlugFox** [comment]
The same issue with `2.0.1` after upgrade from `1.5.3`

```yml
environment:
  sdk: ">=2.18.4 <3.0.0"
  flutter: ">=3.3.8"
  
dependencies:
  injectable: ^2.0.1

dev_dependencies:
  injectable_generator: ^2.0.1
```

**Comment by gabrielaraujoz** [comment]
@Milad-Akarie sorry about that, but I have also got that problem when upgrading. If you have any ideia how I can help, let me know!

**Comment by naamapps** [comment]
@Milad-Akarie I tried replicating this with a simpler setup but without any luck..
I am too confused as to what is the issue here exactly.
In my project I have a pretty complex setup, but with previous version, injectable knew how to handle it.

Will try more and update you if I succeed with replicating the issue.

**Comment by aaumond** [comment]
I have the same problem in my case, I finally downgraded to 1.5.3 version.

**Comment by Threehair** [comment]
Had the same issue. Downgraded to 1.5.3 and it works. No idea what's going on.

For me, the whole problem started when I upgraded autoroute from 5.0.2 to 5.0.4. (I also did backtracked to 5.0.2 just in case -- although I have no idea if the autoroute upgrade has anything to do with the injectable issue.)

**Comment by naamapps** [comment]
@Milad-Akarie I can't seem to replicate with a simple setup..
Do you updates maybe on a solution?

**Comment by Milad-Akarie** [comment]
@naamapps  I just pushed a fix to a barrel files related issue `injectable_generator: 2.1.1 `
can you check if it fixes the registration order issue?

**Comment by Milad-Akarie** [comment]
@naamapps  it should be fix in injectable_generator: 2.1.2
